// jshtml/lib/jshtml/Parser.js
/*!
 * JsHtml
 * Copyright(c) 2012 Elmer Bulthuis <elmerbulthuis@gmail.com>
 * MIT Licensed
 */


var util = require("util");
var assert = require("assert");
var RootContext = require('./RootContext');
var tools = require('../tools');

var newline = "\n";


var base = require("2kenizer");
module.exports = (function(target){

	util.inherits(target, base);

	target.prototype.onToken = function onToken(token, buffer) {
		/*
		if the debug option is on, track the line number
		*/
		for(var index = buffer.indexOf(newline); ~index && index < token.index; index = buffer.indexOf(newline, index + newline.length))	{
			this.line++;
		}

		this.options.trace && console.log(this.line, token.category, token[0]);
		this.context.content += buffer;
		this.context.echo(buffer, {useFilter:true, atEnd: token.category === null});

		do{
			var previousContext = this.context;
			token.redo = false;
			this.context = this.context.onToken(token);
			if(!this.context && this.token.category) throw new SyntaxError("Parse error");
			this.categories = this.context.categories;
			this.context.line = this.line;
			if(this.context === previousContext) this.context.content += token[0];
		} while(token.redo);

		/*
		if there is no token, we are finished
		*/
		if(!token.category && this.context !== this){
			throw new Error("parse error");
		}

		for(var index = token[0].indexOf(newline); ~index && index < token[0].length; index = token[0].indexOf(newline, index + newline.length))	{
			this.line++;
		}
		
	};//onToken


	return target;

})(function(echo, options){
	this.options = tools.extend({
		anchor: "@"
		, textTag: "text"
		, debug: false
		, trace: false
	}, options);

	//var regExpAnchor = tools.regExpEncode(this.options.anchor);
	var expressions = {
		//"anchor":	new RegExp("(\\w*)" + regExpAnchor + "((?:" + regExpAnchor + ")*)(?=[^])")
		"anchor":	this.options.anchor

		, "default":	null
		, "whitespace":	/\s+/
		, "semicolon":	/(\s*)(;)/		

		, "jsIdentifier":	/^(\s*)([$_A-Za-z][$_A-Za-z0-9]*)/
		, "jsMember":		/^(\.)([$_A-Za-z][$_A-Za-z0-9]*)/

		, "jsBeginBlock":	"{"
		, "jsEndBlock":	"}"
		
		, "jsBeginGroup":	"("
		, "jsEndGroup":	")"
		
		, "jsBeginArray":	"["
		, "jsEndArray":	"]"

		, "jsDoubleQuote":	/"(?:\\.|.)*?[^\\]?"/
		, "jsSingleQuote":	/'(?:\\.|.)*?[^\\]?'/
		, "jsRegExp":		/\/(?:\\.|.)+?[^\\]?\/[gim]*/

		, "jsCommentLine":	/\/\/(.*?)\n/
		, "jsCommentBlock":	/\/\*([\s\S]*?)\*\//

		, "tagComment":	/<!--([\s\S]*?)-->\s*/
		, "tagBeginOpen":		/<([\w:]+)/
		, "tagBeginClose":		/(\/?>)\s*/
		, "tagEnd":		/<\/([\w:]+)\s*>\s*/
	};

	base.call(this, this.onToken, expressions, this.options);


	this.echo = echo;

	this.line = 1;

	this.context = new RootContext(this, null, null);
	this.categories = this.context.categories;
	this.context.line = this.line;
	this.context.isRoot = true;

});
